integration/setup.go

```diff
@@ -0,0 +1,447 @@
+// Copyright (c) 2018 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package integration
+
+import (
+	"fmt"
+	"strconv"
+	"sync"
+	"testing"
+	"time"
+
+	"github.com/m3db/m3cluster/client"
+	"github.com/m3db/m3cluster/kv"
+	"github.com/m3db/m3cluster/kv/mem"
+	"github.com/m3db/m3cluster/placement"
+	"github.com/m3db/m3cluster/placement/service"
+	"github.com/m3db/m3cluster/placement/storage"
+	"github.com/m3db/m3cluster/services"
+	"github.com/m3db/m3msg/consumer"
+	"github.com/m3db/m3msg/producer"
+	"github.com/m3db/m3msg/producer/buffer"
+	"github.com/m3db/m3msg/producer/writer"
+	"github.com/m3db/m3msg/topic"
+	"github.com/m3db/m3x/pool"
+	"github.com/m3db/m3x/retry"
+
+	"github.com/golang/mock/gomock"
+	"github.com/stretchr/testify/require"
+	"go.uber.org/atomic"
+)
+
+const (
+	numConcurrentMessages = 10
+	numberOfShards        = 10
+	msgPerShard           = 100
+)
+
+type consumerServiceConfig struct {
+	ct        topic.ConsumptionType
+	instances int
+	replicas  int
+}
+
+type op struct {
+	progressPct int
+	fn          func()
+}
+
+func newTestSetup(
+	t *testing.T,
+	ctrl *gomock.Controller,
+	numberOfProducers int,
+	cscs []consumerServiceConfig,
+) *setup {
+	store := mem.NewStore()
+	configService := client.NewMockClient(ctrl)
+	configService.EXPECT().Store(gomock.Any()).Return(store, nil)
```

Comment:
weird to see mocks in integration tests, any chance you can use the embedded etcd from https://github.com/m3db/m3cluster/tree/master/integration/etcd

Reply:
Because I had to mock some Data, which is supposed to be implemented by users of this library, it could be done by creating a dummy implementation but I feel mock can be used to exam the calls of the object as well which is nice.

Since I had to use a mock so I just mocked the client service, the embedded etcd sometimes closes its go routines too late, which makes the leak detection flaky sometimes.